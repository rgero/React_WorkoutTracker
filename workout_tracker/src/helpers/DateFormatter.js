export default (targetDate) => {

    if (typeof(targetDate) === 'string' || targetDate instanceof String)
    {
        targetDate = new Date(targetDate);
    }
    let formattedDate = targetDate;
    formattedDate = new Date(formattedDate.getTime());
    return formattedDate.toISOString().split('T')[0];
}