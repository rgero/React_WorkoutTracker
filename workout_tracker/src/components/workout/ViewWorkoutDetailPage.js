import React, { useContext, useEffect, useState } from 'react';
import {useParams} from "react-router-dom";
import Container from 'react-bootstrap/Container';

import {Context as WorkoutContext} from '../../context/WorkoutContext';

import DateFormatter from '../../helpers/DateFormatter';

const ViewWorkoutDetailsPage = ()=> 
{
    const {state, fetchWorkoutByID} = useContext(WorkoutContext);
    const [isLoaded, setLoaded] = useState(false);
    const { id } = useParams();

    useEffect(()=> {
        const GetWorkoutByID = async (id) => {
            await fetchWorkoutByID(id);
            setLoaded(true);
        }
        GetWorkoutByID(id);
    }, []);

    if (state.workoutDate)
    {
        return (
            <Container fluid="md">
                <h1>{DateFormatter(state.workoutDate)}</h1>
            </Container>
            )
    } else {
        return (<div>Loading</div>)
    }
}

export default ViewWorkoutDetailsPage;