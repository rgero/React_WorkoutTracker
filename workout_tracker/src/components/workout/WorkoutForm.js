import React, {useState, useEffect} from 'react';

import Form from 'react-bootstrap/Form';
import InputGroup from 'react-bootstrap/InputGroup';
import Container from 'react-bootstrap/Container';
import Button from 'react-bootstrap/Button';

import '../../styles/components/inputs.css'
import '../../styles/components/WorkoutForm.css';

import { ExerciseList } from '../exercise/ExerciseList';
import { ExerciseForm } from '../exercise/ExerciseForm';
import DateFormatter from '../../helpers/DateFormatter';

export const WorkoutForm = ({workout = {}, errorMessage, onSubmit})=> 
{
    const [workoutDate, setWorkoutDate] = useState(workout.workoutDate ? DateFormatter(new Date(workout.workoutDate)) : DateFormatter(new Date()));
    const [exerciseList, setExerciseList] = useState(workout.exerciseList ? workout.exerciseList : []);
    const [notes, setNotes] = useState(workout.notes ? workout.notes : "");
    const [error, setError] = useState("");

    const addExercise = (exercise) => {
        setExerciseList([...exerciseList, exercise]);
    }
    
    const deleteExercise = (index) => {
        let newExerciseList = exerciseList;
        newExerciseList.splice(index, 1);
        setExerciseList( [...newExerciseList] );
    }

    const submitWorkout = (e) => {
        e.preventDefault();
        if (!workoutDate)
        {
            setError("Need a workout date");
        }
        if (exerciseList.length < 1)
        {
            setError("Need at least one exercise");
        }
        if (error)
        {
            return;
        }
        let workout = {
            workoutDate,
            exerciseList,
            notes
        }

        onSubmit(workout);
    }

    return (
        <Container fluid="md">
            <Form id="addWorkout" onSubmit={submitWorkout}>
                <InputGroup className="mb-3">
                    <InputGroup.Text>Date</InputGroup.Text>
                    <Form.Control type="date" value={workoutDate} onChange={e => setWorkoutDate(e.target.value)}/>
                </InputGroup>
                <InputGroup className="mb-3">
                    <InputGroup.Text>Notes</InputGroup.Text>
                    <Form.Control as="textarea" value={notes} rows={3} onChange={e => setNotes(e.target.value)} />
                </InputGroup>
            </Form>
            <div className="exerciseForm">
                <ExerciseForm onSubmit={addExercise}/>
            </div>
            <div className="exerciseList">
                <ExerciseList exerciseList={exerciseList} onDelete={deleteExercise}/>
            </div>
            <Button form="addWorkout" variant="outline-secondary" type="submit">
                    Submit Workout
            </Button>
        </Container>

    )
}