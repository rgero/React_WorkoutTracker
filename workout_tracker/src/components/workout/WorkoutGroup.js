import React from 'react';
import Accordion from 'react-bootstrap/Accordion';
import { WorkoutListItem } from "./WorkoutListItem";
import { GetMonthName, GetOrderedValues } from '../../helpers/WorkoutListProcesser';

const WorkoutGroup = ({workoutList, descendingOrder}) => {
    let months = GetOrderedValues(workoutList, descendingOrder);
    return (
        <div>
            {
                months.map((month, index)=> (
                    <Accordion defaultActiveKey={index} key={`WorkoutGroup_${index}`}>   
                        <Accordion.Item>
                            <Accordion.Header>{GetMonthName(month)}</Accordion.Header>
                            <Accordion.Body>
                                {
                                    workoutList[month].map((workout, index) => (
                                            <WorkoutListItem workout={workout} key={`WorkoutListItem_${index}`} />
                                        ))
                                }
                            </Accordion.Body>
                        </Accordion.Item>
                    </Accordion>
                ))
            }
        </div>
      );
}

export default WorkoutGroup;