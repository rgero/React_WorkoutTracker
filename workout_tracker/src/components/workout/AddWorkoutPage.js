import React, { useContext, } from 'react';
import Container from 'react-bootstrap/Container';

import {Context as WorkoutContext} from '../../context/WorkoutContext';
import { WorkoutForm } from './WorkoutForm';

export const AddWorkoutPage = ()=> 
{
    const {state, createWorkout} = useContext(WorkoutContext);
    return (
        <Container fluid="md">
            <h1>Create a new workout</h1>
            <WorkoutForm
                errorMessage={state.errorMessage}
                onSubmit={({workout})=> createWorkout({workout})}
            />
        </Container>
    )
}