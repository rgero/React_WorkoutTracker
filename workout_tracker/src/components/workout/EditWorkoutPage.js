import React, { useContext, useEffect, useState } from 'react';
import {useParams} from "react-router-dom";
import Container from 'react-bootstrap/Container';

import {Context as WorkoutContext} from '../../context/WorkoutContext';
import { WorkoutForm } from './WorkoutForm';

export const EditWorkoutPage = ( )=> 
{
    const {state, fetchWorkoutByID, updateWorkout} = useContext(WorkoutContext);
    const [isLoaded, setLoaded] = useState(false);
    const { id } = useParams();

    useEffect(()=> {
        const GetWorkoutByID = async (id) => {
            await fetchWorkoutByID(id);
            setLoaded(true);
        }
        GetWorkoutByID(id);
    }, []);

    if (state.workoutDate)
    {
        return (
            <Container fluid="md">
                <h1>Edit a workout</h1>
                <WorkoutForm
                    workout={state}
                    errorMessage={state.errorMessage}
                    onSubmit={(workout)=> updateWorkout(workout)}
                />
            </Container>
            )
    } else {
        return (<div>Loading</div>)
    }
}