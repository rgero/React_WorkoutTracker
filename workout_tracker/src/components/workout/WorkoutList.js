import React from "react";
import Container from "react-bootstrap/Container";

import Collapsible from "react-collapsible";
import {GetOrderedValues, ProcessWorkoutList} from "../../helpers/WorkoutListProcesser";
import WorkoutGroup from "./WorkoutGroup";


export const WorkoutList = ({workoutList=[]}) => 
{

    let organizedList = ProcessWorkoutList(workoutList);
    let yearOrder = GetOrderedValues(organizedList, true);
    return (
        <Container>
            { 
                Object.keys(organizedList).length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>No Workouts</span>
                    </div>
                ) : (
                    <div>
                        
                        {
                            yearOrder.map((year, index)=> (
                                <Collapsible trigger={year}>
                                    <WorkoutGroup {...organizedList[year]} />
                                </Collapsible>
                            ))
                        }
                    </div>
                )
            }
        </Container>
    )
}