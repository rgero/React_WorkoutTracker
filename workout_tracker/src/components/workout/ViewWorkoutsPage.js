import React, {useContext, useEffect} from 'react';
import Container from 'react-bootstrap/Container';
import { WorkoutList } from './WorkoutList';
import {Context as WorkoutContext} from '../../context/WorkoutContext';

export const ViewWorkoutsPage = ()=> 
{
    const {state, fetchWorkouts} = useContext(WorkoutContext);

    useEffect(()=> {
        const processWorkouts = async () => {
            await fetchWorkouts();
        }
       processWorkouts();
    }, []);

    return (
        <Container>
            <div>Workouts</div>
            <WorkoutList workoutList={state} />
        </Container>
    )
}