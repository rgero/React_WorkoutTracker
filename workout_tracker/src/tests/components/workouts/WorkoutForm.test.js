import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow'

import {render, screen} from '@testing-library/react';
import '@testing-library/jest-dom';
import user from '@testing-library/user-event';
import { act } from 'react-dom/test-utils';

import { WorkoutForm } from '../../../components/workout/WorkoutForm';
import { TestExercises, TestWorkouts } from '../../testData';

import DateFormatter from '../../../helpers/DateFormatter';

jest.mock('../../../components/exercise/ExerciseForm', () => ({
    ExerciseForm: ()=> {
        return (<div>Exercise Form Rendered Here</div>);
    }
}));

const mockedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedNavigate
}))

describe("Workout Form Rendering", ()=> {

    test("No Data, matches snapshot", ()=> {
        let workoutInfo = {
            workoutDate: "2022-02-22"
        }
        const pageRender = new ShallowRenderer().render(
            <WorkoutForm workout={workoutInfo}/>
        );
        expect(pageRender).toMatchSnapshot();
    })

    test("Has Data, matches snapshot", ()=> {
        const pageRender = new ShallowRenderer().render(
            <WorkoutForm workout={TestWorkouts[0]}/>
        );
        expect(pageRender).toMatchSnapshot();
    })
})

// Error Handling
// Test fails for no workout date
// Test fails for no exercises
describe("Workout Form", ()=> {

    let workoutDate, workoutNotes;
    let workoutSubmit;

    let mockSubmit = jest.fn();

    let testWorkout = {
        exerciseList: TestExercises
    }

    beforeEach(()=> {

        render(<WorkoutForm workout={testWorkout} onSubmit={mockSubmit} />)

        act(()=> {
            workoutDate = screen.getByLabelText(/workoutdate/i);
            workoutNotes = screen.getByRole("textbox", {name: /workoutnotes/i});

            workoutSubmit = screen.getByRole("button", {name: /workoutsubmit/i});
        })

    })

    test("Full form submission", ()=> {
        act(()=> {
            user.click(workoutDate);
            user.keyboard("2022-02-22");

            user.click(workoutNotes);
            user.keyboard("Today was a productive day");

            user.click(workoutSubmit);
        })

        let expectedResult = {
            "_id": null,
            workoutDate: "2022-02-22",
            notes: "Today was a productive day",
            exerciseList: TestExercises
        }

        expect(mockSubmit).toHaveBeenCalled();
        expect(mockSubmit).toHaveBeenCalledWith(expectedResult);
    })

    test("Min form submission", ()=> {
        act(()=> {
            user.click(workoutDate);
            user.keyboard("2022-02-01");

            user.click(workoutSubmit);
        })

        let expectedResult = {
            "_id": null,
            workoutDate: "2022-02-01",
            notes: "",
            exerciseList: TestExercises
        }

        expect(mockSubmit).toHaveBeenCalled();
        expect(mockSubmit).toHaveBeenCalledWith(expectedResult);
    })

    test("Date automatically filled", ()=> {
        act(()=> {
            user.click(workoutSubmit);
        })

        let expectedResult = {
            "_id": null,
            workoutDate: DateFormatter(new Date()),
            notes: "",
            exerciseList: TestExercises
        }

        expect(mockSubmit).toHaveBeenCalled();
        expect(mockSubmit).toHaveBeenCalledWith(expectedResult);
    })
})