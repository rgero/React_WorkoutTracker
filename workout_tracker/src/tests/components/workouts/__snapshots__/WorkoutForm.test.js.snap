// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workout Form Rendering Has Data, matches snapshot 1`] = `
<Container
  fluid="md"
>
  <Row
    className="justify-content-md-center"
  >
    <Col
      md="8"
    >
      <Card>
        <CardHeader>
          <h4>
            Exercise Information
          </h4>
        </CardHeader>
        <CardBody>
          <Form
            id="addWorkout"
            onSubmit={[Function]}
          >
            <FloatingLabel
              className="mb-3"
              controlId="floatingWorkoutDate"
              label="Workout Date"
            >
              <FormControl
                aria-label="workoutDate"
                onChange={[Function]}
                type="date"
                value={"2023-03-08T00:00:00.000Z"}
              />
            </FloatingLabel>
            <FloatingLabel
              className="mb-3"
              controlId="floatingNotes"
              label="Notes"
            >
              <FormControl
                aria-label="workoutNotes"
                as="textarea"
                onChange={[Function]}
                rows={3}
                value="Testing Notes for Testing"
              />
            </FloatingLabel>
          </Form>
        </CardBody>
      </Card>
      <Card>
        <CardHeader>
          <h4>
            Add/Edit Exercises
          </h4>
        </CardHeader>
        <CardBody>
          <ExerciseForm
            onSubmit={[Function]}
          />
        </CardBody>
      </Card>
      <Accordion
        defaultActiveKey="1"
      >
        <AccordionItem>
          <AccordionHeader>
            Exercise List
          </AccordionHeader>
          <AccordionBody>
            <ExerciseList
              exerciseList={
                Array [
                  Object {
                    "muscleGroup": "Chest",
                    "name": "Bench Press",
                    "notes": "Wrist is feeling a bit sore",
                    "setList": Array [
                      Object {
                        "reps": 20,
                        "weight": 275,
                      },
                      Object {
                        "reps": 20,
                        "weight": 300,
                      },
                      Object {
                        "reps": 3,
                        "weight": 315,
                      },
                    ],
                  },
                  Object {
                    "muscleGroup": "Shoulders",
                    "name": "Shrugs",
                    "notes": "Dumbbells",
                    "setList": Array [
                      Object {
                        "reps": 40,
                        "weight": 100,
                      },
                      Object {
                        "reps": 40,
                        "weight": 100,
                      },
                      Object {
                        "reps": 40,
                        "weight": 100,
                      },
                      Object {
                        "reps": 40,
                        "weight": 100,
                      },
                    ],
                  },
                ]
              }
              onDelete={[Function]}
            />
          </AccordionBody>
        </AccordionItem>
      </Accordion>
      <Button
        aria-label="workoutSubmit"
        form="addWorkout"
        type="submit"
        variant="outline-secondary"
      >
        Submit Workout
      </Button>
    </Col>
  </Row>
</Container>
`;

exports[`Workout Form Rendering No Data, matches snapshot 1`] = `
<Container
  fluid="md"
>
  <Row
    className="justify-content-md-center"
  >
    <Col
      md="8"
    >
      <Card>
        <CardHeader>
          <h4>
            Exercise Information
          </h4>
        </CardHeader>
        <CardBody>
          <Form
            id="addWorkout"
            onSubmit={[Function]}
          >
            <FloatingLabel
              className="mb-3"
              controlId="floatingWorkoutDate"
              label="Workout Date"
            >
              <FormControl
                aria-label="workoutDate"
                onChange={[Function]}
                type="date"
                value={"2022-02-22T05:00:00.000Z"}
              />
            </FloatingLabel>
            <FloatingLabel
              className="mb-3"
              controlId="floatingNotes"
              label="Notes"
            >
              <FormControl
                aria-label="workoutNotes"
                as="textarea"
                onChange={[Function]}
                rows={3}
                value=""
              />
            </FloatingLabel>
          </Form>
        </CardBody>
      </Card>
      <Card>
        <CardHeader>
          <h4>
            Add/Edit Exercises
          </h4>
        </CardHeader>
        <CardBody>
          <ExerciseForm
            onSubmit={[Function]}
          />
        </CardBody>
      </Card>
      <Button
        aria-label="workoutSubmit"
        form="addWorkout"
        type="submit"
        variant="outline-secondary"
      >
        Submit Workout
      </Button>
    </Col>
  </Row>
</Container>
`;
